### PR TITLE
[wasm] detect WASM EXIT on v8 and nodejs

### PR DESCRIPTION
**Pull request by pavelsavara**
Makes detection of https://github.com/dotnet/runtime/issues/85675 easier

also fixes https://github.com/dotnet/xharness/issues/920

### CODE REVIEW

**Review by lewing — APPROVED** [review]
🤞

**Comment by lewing** [comment]
Let's hope this doesn't set everything on fire

**Comment by pavelsavara** [comment]
I hope that we would see the fire on the dependency update PR first

File: src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/JS/WasmTestCommand.cs [hunk 1]
```diff
@@ -177,6 +179,11 @@ protected override async Task<ExitCode> InvokeInternal(ILogger logger)
             }
             else
             {
+                if(!wasmExitReceivedTcs.Task.IsCompletedSuccessfully)
+                {
+                    logger.LogError("Application exited without sending a wasm exit message.");
+                    return ExitCode.RETURN_CODE_NOT_SET;
+                }
```
Comment (radical) [inline src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/JS/WasmTestCommand.cs:186]: This should be before line 175, as it shouldn't depend on what `result.ExitCode` is.

Comment (pavelsavara) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/JS/WasmTestCommand.cs:186]: If v8 failed for any other reason, for example that test-main.js doesn't exist, I don't want it to be covered by message about WASM EXIT

**Comment by premun** [comment]
It might be beneficial to invest into updating the E2E tests for WASM in this repo so that you don't need to postpone testing for runtime dependency PRs: https://github.com/dotnet/xharness/issues/920

I can merge this PR for you through the timeouts in the Apple queues but the WASM leg is failing and that one should probably go green first.

File: tests/integration-tests/WASM/WASM.Helix.SDK.Tests.proj [hunk 2]
```diff
@@ -18,7 +18,7 @@
 
     <ItemGroup>
       <HelixWorkItem Include="System.Reflection.Extensions.Tests">
```
Comment (premun) [inline tests/integration-tests/WASM/WASM.Helix.SDK.Tests.proj:20]:
```suggestion
      <HelixWorkItem Include="Common.Tests">
```

**Comment by pavelsavara** [comment]
`can't cd to /datadisks/disk1/work/C4620A42/w/A473091C/u/Common.Tests`

**Comment by premun** [comment]
@pavelsavara I didn't check your payload really, only noticed the mismatch in the name of the workitem so I fixed that. I can put a root-level directory in the zip that is uploaded.

File: tests/integration-tests/WASM/WASM.Helix.SDK.Tests.proj [hunk 3]
```diff
@@ -17,7 +17,7 @@
     <Message Text="Downloaded @(TestPayloadArchive) for XHarness Test purposes" Importance="High" />
 
     <ItemGroup>
-      <HelixWorkItem Include="System.Reflection.Extensions.Tests">
+      <HelixWorkItem Include="Common.Tests">
         <Command>cd "$HELIX_WORKITEM_PAYLOAD/Common.Tests" &amp;&amp; dotnet exec $XHARNESS_CLI_PATH wasm test --engine=V8 --js-file=test-main.js -v --output-directory="$HELIX_WORKITEM_UPLOAD_ROOT" -- --run WasmTestRunner.dll Common.Tests.dll -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing</Command>
```
Comment (premun) [inline tests/integration-tests/WASM/WASM.Helix.SDK.Tests.proj:21]: @pavelsavara or we can do this:
```suggestion
<Command>dotnet exec "$XHARNESS_CLI_PATH" wasm test --engine=V8 --js-file=test-main.js -v --output-directory="$HELIX_WORKITEM_UPLOAD_ROOT" -- --run WasmTestRunner.dll Common.Tests.dll -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing</Command>
```

**Comment by pavelsavara** [comment]
`info: V8 version: 8.5.183` this is very very old version of V8, could we upgrade ?

**Comment by radical** [comment]
@pavelsavara is there a runtime issue that has details of what is being fixed here?

**Comment by pavelsavara** [comment]
> @pavelsavara is there a runtime issue that has details of what is being fixed here?

I created new issue for it. 
https://github.com/dotnet/runtime/issues/85675
This PR makes detection possible. 

I hope I have draft of fix on runtime repo https://github.com/dotnet/runtime/pull/85542

**Comment by premun** [comment]
> `info: V8 version: 8.5.183` this is very very old version of V8, could we upgrade ?
> 
> Ideally to latest, I have `11.5.3`

@pavelsavara I guess we can update that, not sure if anyone else depends on that. I logged https://github.com/dotnet/arcade/issues/13399 and ETA would be next Wednesday.

**Comment by pavelsavara** [comment]
> @pavelsavara would it better to run this in one of your docker containers? I understand that's how you do it in runtime, right?

eventually yes, but I don't want to invest my time into it right now. Is that easy to switch ?

I think this one would be fine `(Ubuntu.1804.Amd64)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-webassembly`

cc @radekdoulik

**Comment by premun** [comment]
@pavelsavara I think it might be easy enough by setting the `HelixTargetQueue` in [WASM.Helix.SDK.Tests.proj](https://github.com/dotnet/xharness/pull/1014/files#diff-4e23d7bd178bdf12292b6307633daf1bc046a757379546e82f52e78aa3e121d9)

**Comment by radekdoulik** [comment]
[WASM.Helix.SDK.Tests.proj](https://github.com/dotnet/xharness/pull/1014/files#diff-4e23d7bd178bdf12292b6307633daf1bc046a757379546e82f52e78aa3e121d9) currently uses ubuntu 20.04, so you probably want to use `ubuntu-20.04-helix-wasm-amd64` docker image

**Review by radekdoulik — APPROVED** [review]
LGTM besides the old v8 issue

**Comment by pavelsavara** [comment]
I see `<HelixTargetQueue Include="ubuntu.2004.amd64.open"/>`  in `WASM.Helix.SDK.Tests.proj` but in yaml we have name with some namespaces and prefix in the bracket. I don't know what they are. I will go with naive approach.

**Comment by premun** [comment]
@pavelsavara the notation is described here - https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.Helix/Sdk#docker-support

**Comment by akoeplinger** [comment]
Should we get this in?

**Comment by pavelsavara** [comment]
> Should we get this in?

thanks for fixing it! I think we should merge it. @radical  ?

**Comment by ViktorHofer** [comment]
I'm seeing the introduce log in https://github.com/dotnet/runtime/pull/96094:

https://[helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-96094-merge-0a54c16d6a9f4828a8/normal-System.Runtime.Tests/1/console.0ee49343.log?helixlogtype=result](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-96094-merge-0a54c16d6a9f4828a8/normal-System.Runtime.Tests/1/console.0ee49343.log?helixlogtype=result)

> [14:53:17] info: Waiting to flush log messages with a timeout of 120 secs ..
[14:53:17] info: Received expected 17558878 of /root/helix/work/workitem/uploads/xharness-output/testResults.xml
[14:53:17] info: Finished writing 17558878 bytes of RESULTXML
[14:53:17] info: Xml file was written to the provided writer.
[14:53:17] info: Tests run: 63620 Passed: 63518 Inconclusive: 0 Failed: 0 Ignored: 0 Skipped: 102
[14:53:17] info: test-main.js exiting WasmTestRunner.dll System.Runtime.Tests.dll with result 0
[14:53:17] info: console.info: MONO_WASM: forceDisposeProxies done: 0 imports, 0 exports, 0 GCHandles, 0 JSHandles.
[14:53:17] fail: Application exited without sending a wasm exit message.
XHarness exit code: 82 (RETURN_CODE_NOT_SET)
/root/helix/work/workitem/e /root/helix/work/workitem/e
----- end Mon Dec 18 02:53:17 PM UTC 2023 ----- exit code 82 ----------------------------------------------------------
XHarness artifacts: /root/helix/work/workitem/uploads/xharness-output

**Comment by pavelsavara** [comment]
This is the same test on runtime main [Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-95959-merge-619cc21f67a74cfcaa/normal-System.Runtime.Tests/1/console.b54bba60.log?helixlogtype=result)

It also didn't log exit. And this was reason why we wanted this detection. 

But we perhaps need to not set `RETURN_CODE_NOT_SET` until we resolve existing problems on runtime repo.

I will create another PR to make this softer.

**Pull request merged**